File: pattern-controller.js
var config = require('./config');
var pattern_lib = require('./patterns');
var clone = require('clone');
var Pattern = require('./patterns/pattern');

var Controller = function(settings)
{
	var self = this;
	this.emitter = settings.emitter;
	this.lanterns = settings.lanterns;

	this.timer = 0;
    this.last_run = 0;
    this.current_pattern = null;

	this.emitter.on('newConnection', function(){
		self.sendPatternNames();
	});

	this.emitter.on('controllerChangePattern', function(data){
		var duration = Math.randomInt(config.patterns.min, config.patterns.max);
		var pattern = new pattern_lib.patterns[data.pattern]();
		self.setPattern(data.channel, pattern, duration);
	});
}

Controller.prototype.sendPatternNames = function(){
	var names = pattern_lib.group.names;
	this.emitter.emit('patternNames', names);
}

Controller.prototype.tick = function()
{
	if(this.current_pattern instanceof Pattern)
	{
		this.current_pattern.tick();
	}

	var elapsed_time = Date.now() - this.last_run;
	if(elapsed_time > 1000)
	{
		this.timer--;
		this.last_run = Date.now();
	}

	if(this.timer <= 0)
	{
		this.setGroupPattern();
	}
}

Controller.prototype.setGroupPattern = function()
{
    var self = this;
	var duration = Math.randomInt(config.patterns.min, config.patterns.max);
	var pattern_num  = Math.randomInt(0, pattern_lib.group.patterns.length - 1);

    if(config.debug.enabled && !isNaN(parseInt(config.debug.restrict_group_pattern)))
    {
        pattern_num = config.debug.restrict_group_pattern;
    }
	
	if(this.current_pattern)
	{
		this.current_pattern.removeAllListeners('render').removeAllListeners('getColor');
	}

	this.current_pattern = new pattern_lib.group.patterns[pattern_num]();
	
    this.current_pattern.init();

	this.current_pattern.on('render', this.render.bind(this));
	this.current_pattern.on('getColor', function(i){
		self.current_pattern.emit('pixelValue', self.getPixelValue(i));
	});

    this.timer = duration;

	this.emitter.emit('updateGroupPattern', true);
	console.debug('change pattern to %s for duration %s', pattern_lib.group.names[pattern_num], duration);
}

Controller.prototype.render = function()
{
    for(lantern in this.lanterns)
    {
        this.lanterns[lantern].mapPixels(this.current_pattern.shader.bind(this.current_pattern));
    }
}

module.exports = Controller;

File: patternController.py
import time
import random
import logging

import patterns

logger = logging.getLogger('iopener')

class patternController(object):
	def __init__(self, lanterns = []):
		self.time = 0 

		self.lanterns = lanterns

		self.timers = [0 for l in self.lanterns]
		
		self.group_timer = 100

		random.seed()
	
	def tick(self, dt):
		self.time += dt
		if(self.time >= 1):
			self.update()
			self.time -= 1.0
		

	def update(self):
		for i, timer in enumerate(self.timers):
			self.timers[i] -= 1

			if self.timers[i] <= 0:
				self.changePattern(i)

	def changePattern(self, i):
		pattern_class, pattern_name = patterns.randomPattern()
		pattern = pattern_class(channel = self.lanterns[i].channel)
		duration = random.randint(20, 40)

		self.lanterns[i].setPattern(pattern)
		self.timers[i] = duration 

		logger.debug("change pattern for lantern %s to pattern %s for %s" % (self.lanterns[i].channel, pattern_name, duration))
